:mortar_board: Topic instructions --- Instruction set

jameshughes89 opened the pull request:
### What

Discuss what instructions are at a high-level, and describe 13 instructions for the system in its current state

### Testing

:+1: 

### Additional Notes

Note that the motivation for labeling addition as `ADDB` (and similarly for `SUBB`) is as follows. 

`ADDB` means "Add the contents of register B to the contents of register A". This means that if `A=2` and `B=3`, after the operation `A=5`. One could argue that `ADDA` makes more sense as it's adding something to A. However, A is serving as an "accumulator", a common register type, and the idea is that `B` is being added to something. If the system had another register called `C`, `ADDC` could mean "add the contents of C to A".  

The only area where it gets harry is, what if I added an instruction like `ADDR` to add the contents of `A` and `B` and store the result in some specified RAM address. If I do that, the `R` is the destination, but the `B` in `ADDB` is the source. This is somewhat confusing. That said, (a) this is a _what if_ and (b) if I did add this instruction, I think I could call it something like `SAVS` for like, save sum, to follow a similar pattern to the already existing (action/source)

jameshughes89 committed:
get ideas down

jameshughes89 committed:
explain instruction set

jameshughes89 committed:
add table

jameshughes89 committed:
start the instruction description

jameshughes89 committed:
load Bs

jameshughes89 committed:
save

jameshughes89 committed:
do add and sub

jameshughes89 committed:
jump and noops

jameshughes89 committed:
write out variations

jameshughes89 committed:
hald

jameshughes89 committed:
clean

jameshughes89 commented on site/topics/control-logic/instructions-microcodes.rst:
```diff
@@ -189,6 +189,282 @@ Fetch and Instruction Register
+* Following the table is a description of each instruction
+
+
+.. list-table:: An Instruction Set for the Current ESAP System
```
![image](https://github.com/user-attachments/assets/74f20573-5b35-4c1e-95e9-27c0ce23aacf)

twentylemon commented:
> `ADDB` means "Add the contents of register B to the contents of register A". This means that if `A=2` and `B=3`, after the operation `A=5`.

Alternatively, `ADXY` signifying `x += y`. Both registers are shown, the operator, and the order-ness of it. Works similarly for `SBXY`, `MLXY`, `DVXY`, `EXXP` or `PWXY`, etc.

twentylemon commented:
Or, further, these are just instruction words. They don't need to be four characters, they just need to map to the proper opcode. `ADDXY`, or `ADDRX` seem fine?

jameshughes89 commented:
> Or, further, these are just instruction words. They don't need to be four characters, they just need to map to the proper opcode. `ADDXY`, or `ADDRX` seem fine?

Although true, I prefer keeping it to a four for consistency and visual clarity.

jameshughes89 commented:
> Or, further, these are just instruction words. They don't need to be four characters, they just need to map to the proper opcode. `ADDXY`, or `ADDRX` seem fine?

I like this idea a lot, but with the change to `ADXY -> ADAB`. 

Leaves more room for an alternative instruction where adding takes an operand and saves the result in RAM somewhere (though, idk what that would be called).

jameshughes89 committed:
change ADDB & SUBB -> ADAB & SUAB

jameshughes89 commented:
Updated:
- `ADDB -> ADAB`
- `SUBB -> SUAB`

Also updated the descriptions a little

This will require updating the tags in #174 and a new PR to update the comments in the code for the control logic.

twentylemon commented on site/topics/control-logic/instructions-microcodes.rst:
```diff
@@ -189,6 +189,282 @@ Fetch and Instruction Register
+    * Assuming the hardware is sufficient to perform such an instruction
+
+
+* Below are 13 instructions that can be included on the system
```
More natural flow, imo, would be to note desirable characteristics of an instruction set upfront and motivate them. Lead into the instructions rather than assert them.

jameshughes89 replied:
Good clal

twentylemon reviewed (commented):
Partial. I gotta eat.

jameshughes89 committed:
disuss functionality

jameshughes89 committed:
Merge branch 'main' into topic-instructions-instruction_set
